swig/ChunkedArray.hpp

```diff
@@ -0,0 +1,240 @@
+        if (this->empty()) {
+            return;
+        }
+
+        const size_t full_chunks = this->get_chunks_count() - 1;
+
+        // Copy full chunks:
+        size_t i = 0;
+        for(size_t chunk = 0; chunk < full_chunks; ++chunk) {
+            T* chunk_ptr = _chunks[chunk];
+            for(size_t chunk_pos = 0; chunk_pos < _chunk_size; ++chunk_pos) {
+                other[i++] = chunk_ptr[chunk_pos];
+            }
+        }
+        // Copy filled values from last chunk only:
+        const size_t last_chunk_elems = this->get_current_chunk_added_count();
+        T* chunk_ptr = _chunks[full_chunks];
+        for(size_t chunk_pos = 0; chunk_pos < last_chunk_elems; ++chunk_pos) {
+            other[i++] = chunk_ptr[chunk_pos];
+        }
+    }
+
+    /**
+     * Return value from array of chunks.
+     *
+     * @param chunks_index index of the chunk
+     * @param index index within chunk
+     * @param on_fail_value sentinel value. If out of bounds returns that value.
+     *
+     * @return pointer or nullptr if index is out of bounds.
+     */
+    T getitem(size_t chunks_index, size_t index, T on_fail_value) noexcept
+    {
+        if (within_bounds(chunks_index, index))
+            return _chunks[chunks_index][index];
+        else
+            return on_fail_value;
+    }
+
+    /**
+     *
+     * @param chunks_index index of the chunk
+     * @param index index within chunk
+     * @param value value to store
+     *
+     * @return 0 = success, -1 = out of bounds access.
+     */
+    int setitem(size_t chunks_index, size_t index, T value) noexcept
+    {
+        if (within_bounds(chunks_index, index))
+        {
+            _chunks[chunks_index][index] = value;
+            return 0;
+        } else {
+            return -1;
+        }
+    }
+
+    /**
+     * To reset storage call this.
+     * Will release existing resources and prepare for reuse.
+     */
+    void clear() noexcept
+    {
+        release();
+        new_chunk();
+    }
+
+    /**
+     * Returns true if is empty.
+     */
+    bool empty() const noexcept
+    {
+        return get_current_chunk_added_count() == 0;
+    }
+
+    /**
+     * Deletes all the allocated chunks.
+     * Do not use container after this! See ``clear()`` instead.
+     */
+    void release() noexcept
+    {
+        std::for_each(_chunks.begin(), _chunks.end(), [](T* c) { delete[] c; });
+        _chunks.clear();
+        _chunks.shrink_to_fit();
+        _current_chunks_idx = 0;
+        _current_chunk_idx = 0;
+    }
+
+    inline bool within_bounds(size_t chunks_index, size_t index) {
+        return (chunks_index < _chunks.size()) && (index < _chunk_size);
+    }
+
+    /**
+     * Adds a new chunk to the array of chunks. Not thread-safe.
+     */
+    void new_chunk()
+    {
+        _chunks.push_back(new (std::nothrow) T[_chunk_size]);
+
+        // Check memory allocation success:
+        if (! _chunks[_chunks.size()-1]) {
```

Comment:
Actually don't think this helps. We'd lose all stored data so far.

Maybe best to let `_chunks.push_back(new T[_chunk_size]);` throw the bad_alloc as usual when memory runs out.

Reply:
However, if we need to expose ChunkedArray in the C API, this needs to be done, and -1/0 returned on the C API as is the case for all other C functions.

I'll let it be for now then.